### ISSUE TITLE
How to use the click that closes a PopUp for next interaction

### ISSUE THREAD

**Issue by pixtur** [ocornut/imgui#3154]
Version: 1.75
Branch: master
Back-ends: imgui.net
Operating System: win10

**My Issue/Question:**
I'm currently testing an interaction in which I use pop-up windows to edit properties of nodes within a graph. Since this is a very frequent interaction, the click to close the popup gets cumbersome and in some instances irritating: because sometimes I'm already looking at another node that I want to select, forgot about the already open popup, and get confused that I have to click twice to selected it.

Is there a way to close the popup without consuming the click?

I believe the primary reason for this behaviour is that I'm checking for...

```
ImGui.IsWindowHovered()
```

I believe I'm looking for a work around that tests for something like "is window visible below mouse even if popup is currently open".

**Screenshots/Video**
![fence-selection](https://user-images.githubusercontent.com/1732545/80079492-a5936680-8550-11ea-9d10-474007f2415a.gif)

**Comment by ocornut** [comment]
`IsWindowHovered(ImGuiHoveredFlags_AllowWhenBlockedByPopup)`

**Comment by ocornut** [comment]
My following question is, can you explain what prevented you to discover this yourself? Is there something we can improve in the header files?

Note this is also what functions like `BeginPopupContextItem()` are using.

I am starting to think that ugly state of cimgui output is a reasonable cause of documentation problems and I am considering rewriting cimgui so it output good-looking files then promote other bindings such as the .net one to allow output reasonable looking headers.

**Comment by pixtur** [comment]
Darn. Again! I'm feeling quite embarrassed for asking stupid questions over and over again.

Your follow up question is totally justified. What I normally do is...

1. Start typing in Rider and see what auto completion shows up. (In this case, the optional parameters for `IsWindowHovered` were not emphasized).
2. Open `imgui.h`, start searching and try to read to relevant comments (I have to admit, that I have a very hard time reading and piecing together this documentation. I personally would find it incredibly helpful to have [build-in documentation the .net-wrapper](https://github.com/mellinoe/ImGui.NET/issues/119), but do to the current formatting of .h file, this is not possible). I this case I read the section on PopUps but didn't spot the section on hover flags).
3. I search the imgui-issues and check 5-10 topics that sound related.
4. I look in the imgui-demo for similar interactions and check the implementation.

I never check imgui.cpp.

I normally spent at least 5 minutes searching before writing a new issue and experiment with various alternatives to solve my problem.

Maybe I'm too much of a lazy designer-programmer that hates reading non-wrapped non-formatted header files. I made several attempts to read the complete .h file (I actually started to copy the header file into a mark-down-file to add formatting), but always gave up.  This header file was a major reason I was hesitant to use ImGui in the first place. But I know many people that say this documentation IS the reason they are using the library. And I'm very happy to convinced me to join them.

For me personally, having a "formatted lazy human readable" documentation for imgui would be huge. But I also see why creating and maintaining such a documentation is out of scope. 

As a side node, I completely read https://lodash.com/docs/ , https://momentjs.com/docs/ and many other documentation before using those library. The ImGui documentation itself is excellent and well written. It's just the presentation.

Maybe I will spend some time on the week-end on another attempt of formatting. That would force me to finally read it completely.

**Comment by ocornut** [comment]
Thanks for the feedback. 

- I'm open and interested in results of good attempts to generate documentation (past attempts I've seen yield poor results ihmo and also it is quite bad when people show up, create a bad auto-
generated doc, never update it but keep it on the web for years).

- I would also be happy if there was a serious effort to write AND maintain docs. Past efforts have come from people who aren't yet using the library and ihmo this is not a good starting point, anyway those people tend to disappear after the first hurdle.  (I realize we can argue that the benefit of imgui-inexperience IS valuable, but I believe the benefit of imgui-experience is MORE valuable and the experienced people can always gather and interpret feedback from inexperienced people, we certainly won't run out of people asking questions).

- The way we use pre C++11 enums is really not ideal as in the C++ version there's no hard connection between the function parameter type and the actual enum, so IDE gets confused, however using completion from the prefix should always work. The bad news is that C++20 is heading toward making enums worse than they are and we may not be able to benefit from C++11 ones nicely if C++20 decide to add aggressive warnings everywhere where they are doing now.

- I wonder how much of it is reliant on the IDE you use. With Visual Studio I can Alt+G over some symbol name (even in a _comment_) and find where it comes from. Jumping around is really fast.

- I would suggest pushing work over to cimgui to actually parse and store comments. It's not going to be 100% perfect on the first try, but even if it does 90% you'll have better wrappers.

- I was about to add a comment near the imgui.h description of Popups and saw the comments I wanted to add is _already_ there. This is how the POPUP sections starts:

```
// Popups, Modals
// The properties of popups windows are:
// - They block normal mouse hovering detection outside them. (*)
// - Unless modal, they can be closed by clicking anywhere outside them, or by pressing ESCAPE.
// - Their visibility state (~bool) is held internally by imgui instead of being held by the programmer as we are used to with regular Begin() calls.
//   User can manipulate the visibility state by calling OpenPopup().
// - We default to use the right mouse (ImGuiMouseButton_Right=1) for the Popup Context functions.
// (*) You can use IsItemHovered(ImGuiHoveredFlags_AllowWhenBlockedByPopup) to bypass it and detect hovering even when normally blocked by a popup.
// Those three properties are connected. The library needs to hold their visibility state because it can close popups at any time.
```

- In Demo > Popups > Context Menu it would behave as you expected (right-click can reopen one menu) so based on this behavior one can technically infer this is possible.

- If you have any actionable suggestion even small ones I'm happy to hear them.

**Comment by pixtur** [comment]
Thanks for your reply. I'm certain you already invested a lot of thought into the topic. I'm not familiar enough with c or c++ so I can't really contribute anything. 

I believe in Rider and Visual Studio there is a gap between the c# wrapper and the initial .h files. Although I can see all type declarations and definitions, all comments are missing. That's why I manually search the .h in parallel.

Last year I didn't a quick look how cimgui parses the header file. But I have have no idea how comment sections (like the one about the PopUps and Modals) could automatically extracted for certain topics.

Anyways. I will try to read the documentation more carefully before asking any more questions. I'm really sorry for wasting your time.
